wweval/R/ms_fig4.R

```diff
@@ -0,0 +1,519 @@
+        horizon == "3 wks" ~ 4,
+        horizon == "4 wks" ~ 5,
+        horizon == "overall" ~ 6
+      )
+    )
+
+  relative_crps <- scores_comb |>
+    dplyr::select(location, forecast_date, model, horizon, fig_order, crps) |>
+    dplyr::filter(!is.na(horizon)) |>
+    tidyr::pivot_wider(
+      names_from = model,
+      values_from = crps,
+      id_cols = c(location, forecast_date, horizon, fig_order)
+    ) |>
+    dplyr::mutate(
+      rel_crps = ww / hosp
+    ) |>
+    dplyr::mutate(
+      horizon = forcats::fct_reorder(horizon, fig_order)
+    )
+
+
+  p <- ggplot(relative_crps, aes(
+    x = rel_crps, y = location, color = horizon,
+    fill = horizon
+  )) +
+    geom_violin(alpha = 0.3) +
+    geom_vline(aes(xintercept = 1), linetype = "dashed") +
+    theme_bw() +
+    theme(
+      axis.text.x = element_text(
+        size = 8, vjust = 1,
+        hjust = 1, angle = 45
+      ),
+      axis.title.x = element_text(size = 12),
+      axis.title.y = element_text(size = 10),
+      plot.title = element_text(
+        size = 10,
+        vjust = 0.5, hjust = 0.5
+      )
+    ) +
+    ggtitle("Distribution of CRPS by location, across forecast dates") +
+    ylab("") +
+    xlab("Relative CRPS, lower is better") +
+    scale_x_continuous(trans = "log")
+
+  return(p)
+}
+
+#' Make figure that stratifies across location and forecast dates
+#'
+#' @param scores A tibble of scores by location, forecast date, date and model,
+#' the ouput of `scoringutils::score()` on samples.
+#' @param horizons_to_show A vector of strings indicating the names of the
+#' `horizon` that we want to show on the plot, must be a subset of
+#' `nowcast`, `1 wk`, `2 wks`,`3 wks`, `4 wks` and `overall`
+#'
+#' @return A ggplot object containing plots of the distribution of relative
+#' CRPS scores across location and forecast dates
+#' @export
+make_fig4_rel_crps_overall <- function(scores,
+                                       horizons_to_show = c(
+                                         "nowcast",
+                                         "1 wk", "4 wks",
+                                         "overall"
+                                       )) {
+  scores_by_horizon <- scores |>
+    data.table::as.data.table() |>
+    scoringutils::summarise_scores(by = c(
+      "forecast_date", "location",
+      "model", "horizon"
+    ))
+
+  scores_overall <- scores |>
+    data.table::as.data.table() |>
+    scoringutils::summarise_scores(by = c(
+      "forecast_date", "location",
+      "model"
+    )) |>
+    dplyr::mutate(horizon = "overall")
+
+  scores_comb <- dplyr::bind_rows(scores_by_horizon, scores_overall) |>
+    dplyr::filter(
+      horizon %in% horizons_to_show
+    ) |>
+    dplyr::mutate(
+      fig_order = dplyr::case_when(
+        horizon == "nowcast" ~ 1,
+        horizon == "1 wk" ~ 2,
+        horizon == "2 wks" ~ 3,
+        horizon == "3 wks" ~ 4,
+        horizon == "4 wks" ~ 5,
+        horizon == "overall" ~ 6
+      )
+    )
```

Comment:
It seems like this operation is happening a lot. Perhaps worth creating a helper function for DRY?

Reply:
I'd add the horizon factor ordering into that function too (i.e. this bit)

```R
    dplyr::mutate(
      horizon = forcats::fct_reorder(horizon, fig_order)
    )

```